vignettes/predict.Rmd

```diff
@@ -0,0 +1,109 @@
+---
+title: "Predicting from a MMRM"
+package: mmrm
+output:
+  rmarkdown::html_document:
+          theme: "spacelab"
+          highlight: "kate"
+          toc: true
+          toc_float: true
+vignette: |
+  %\VignetteIndexEntry{Predicting from a MMRM}
+  %\VignetteEncoding{UTF-8}
+  %\VignetteEngine{knitr::rmarkdown}
+editor_options:
+  chunk_output_type: console
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+library(mmrm)
+```
+
+## Prediction of conditional mean
+
+Let $Y^{(i)}$ be the $n$ outcomes of the $i$-th subject and $X^{(i)}$ the $n\times p$
+matrix of covariates for subject $i$.
+Let further $\epsilon^{(i)}\sim\mathcal{N}(0, \Sigma(X^{(i)},\theta))$ be the vector of
+(correlated) residuals where $\theta$ are the covariance parameters.
+
+$$
+Y^{(i)} \sim X^{(i)} \cdot \beta^\top + \epsilon^{(i)}
+$$
+
+Since residuals can be correlated, potentially existing observed outcomes
+of the same individual can be informative for predicting the unobserved 
+valued of the same individual.
+
+Assume that the data is sorted such that $Y^{(i)}_{j} = y^{(i)}_{j}, j = k+1, k+2, \dots, p$ are
+observed and $Y^{(i)}_{j}, j = 1, 2, \dots, k$ are not.
+The special case of all outcomes being unobserved (new individual) is covered
+with $k=p$.
+
+Let further 
+$$
+\Sigma^{(i)}(\theta) = \begin{pmatrix} \Sigma^{(i)}_{new,new}(\theta) & \Sigma^{(i)}_{new,old}(\theta)\\ \Sigma^{(i)}_{old,new}(\theta) & \Sigma^{(i)}_{old,old}(\theta)\end{pmatrix}
+$$
+
+be a block decomposition where
+$\Sigma^{(i)}_{new,old}(\theta) = \big(\Sigma^{(i)}_{j,l}(\theta)\big)_{j = 1\dots k,\, l = 1\ldots k}$ and
+
+Predictions can then be made based on the conditional distribution
+$$
+Y^{(i)}_{1\ldots k}\,|\,X_i,Y^{(i)}_{k+1\ldots p}=y^{(i)}_{k+1\ldots p}\sim\mathcal{N}(\mu, A)
+$$
+
+with 
+
+$$
+\mu = y^{(i)}_{1\ldots k} +  \Sigma^{(i)}_{new,old}(\theta) \, \big(\Sigma^{(i)}_{old,old}(\theta)\big)^{-1} \big(y^{(i)}_{k+1\ldots p} -  (X^{(i)} \cdot \beta^\top)_{k+1\ldots p}\big)
+$$
+and
+
+$$
+A = \Sigma^{(i)}_{new,new}(\theta) - \Sigma^{(i)}_{old,new}(\theta) \big(\Sigma^{(i)}_{old,old}(\theta)\big)^{-1} \Sigma^{(i)}_{new,old}(\theta)
+$$
+
+For implementing `predict()` $\mu$ is required. 
+
+For `predict(interval = "confidence")` additionally standard errors are required.
+These could be derived using the delta methods since $mu$ is a function of the
+estimated model parameters $\beta,\theta$.
+This would require the Jacobian of $mu(\beta,\theta)$.
+
+For `predict(interval = "prediction")` one would use the diagonal of $A$ instead,
+again the delta method could be used to make upper and lower boundaries reflect
+parameter estimation uncertainty.
```

Comment:
OK seems this is the hard part. So given beta and theta, we have the multivariate normal distribution. I wonder if it would be conceptually and even computationally simpler to draw beta, theta from the joint asymptotic normal distribution, centered around the estimates, and using the joint Hessian for the covariance. Then 2-step Monte Carlo sampling:
1) sample beta, theta, calculate from that mu and A.
2) sample the missing y's from the multivariate normal.
Then derive CI as sample quantiles. 
The sampling and computations are fast, can be done in C++, and we maybe need 1000 to have low MC error, so will be still good user experience.

Reply:
and when thinking about this further, instead of joint Hessian for beta, theta, could also just
a) draw theta from its asymptotic normal (we already have vcov for theta)
b) then draw beta from conditional asymptotic normal (we have that as well already)
we could prototype this in R already, without new code in C++.

Reply:
hmhm, would it not be better to avoid sampling alltogether - I would prefer deterministic algos. We have the capability via TMB - it is just not exposed on the R side because all the cov stuff is hard coded in C++.

Reply:
we can expose `get_covariance(X_i, type, theta)` to R. however, it might also be good that this could be reported through the objective function? so that you can obtain the result directly.

Reply:
Maybe through objective function we could get slower optimization? Maybe good to quickly do a option comparison between separate vs objective function interface

Reply:
we now created a new issue for this in #263